README.md
# AhoCorasearch

This lib is an elixir wrapper of the Rust lib Daachorse. 

## Installation

If [available in Hex](https://hex.pm/docs/publish), the package can be installed
by adding `aho_corasearch` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:aho_corasearch, "~> 0.1.0"}
  ]
end
```

Documentation can be found on [HexDocs](https://hexdocs.pm/aho_corasearch/usage.html).

## Basic Usage

```elixir
patterns = [
  {"search", 0},
  {"keyword", 1},
  {"text", 2},
  {"📚", 2},
  {"🕵️‍♂️", 0},
  {"🗝️", 1}
]
{:ok, tree} = AhoCorasearch.build_tree(patterns)
text = "This is some text(📚) you want to search(🕵️‍♂️) for keywords(🗝️)."
AhoCorasearch.search(tree, text)
```
Result looks like:
```elixir
[
  {13, 17, [2]},
  {18, 22, [3]},
  {36, 42, [0]},
  {43, 59, [4]},
  {65, 72, [1]},
  {74, 81, [5]}
]
```

Could be processed like:
```elixir
keys = %{0 => :search, 1 => :keyword, 2 => :text}
Enum.map(results, fn {start, stop, ids} ->
  %{
    range: start..stop,
    match: :binary.part(text, start, stop - start),
    key: Enum.map(ids, &keys[&1])
  }
end)
```
Which returns:
```elixir
[
  %{match: "text", range: 13..17, key: [:text]},
  %{match: "📚", range: 18..22, key: [:text]},
  %{match: "search", range: 36..42, key: [:search]},
  %{match: "🕵️‍♂️", range: 43..59, key: [:search]},
  %{match: "keyword", range: 65..72, key: [:keyword]},
  %{match: "🗝️", range: 74..81, key: [:keyword]}
]
```

Patterns take the form `[{string: String.t(), id: pos_integer()}, ...]`, where `string` is a string you want to search for, and `id` is any positive integer that you can later use for linking matches. 

More examples and a deeper explanation of settings/options can be found in the HexDocs(or the Usage.md file)

## Credits

The underlying Rust library is Daachorse.

- [Github](https://github.com/daac-tools/daachorse)

- [Crates.io](https://crates.io/crates/daachorse)

- [Docs.rs](https://docs.rs/daachorse/1.0.0/daachorse/)

It cannot be overstated how much of the performance(all of it) and capabilities(all but a few minor things like case insensitivity) of this library are due to the underlying Rust lib, daachorse. So a huge thanks to everyone who worked on that, and for releasing it with a permissive OSS license. This library wouldn't be possible without their hard work. 

Daachorse is released under both Apache 2.0 and MIT license.

Daachorse citation:

```
@article{10.1002/spe.3190,
    author = {Kanda, Shunsuke and Akabe, Koichi and Oda, Yusuke},
    title = {Engineering faster double-array {Aho--Corasick} automata},
    journal = {Software: Practice and Experience},
    volume={53},
    number={6},
    pages={1332--1361},
    year={2023},
    keywords = {Aho–Corasick automata, code optimization, double-array, multiple pattern matching},
    doi = {https://doi.org/10.1002/spe.3190},
    url = {https://onlinelibrary.wiley.com/doi/abs/10.1002/spe.3190},
    eprint = {https://onlinelibrary.wiley.com/doi/pdf/10.1002/spe.3190}
}
```

## License

MIT License

Copyright (c) 2023 Peter Richards

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

CHANGELOG.md
# CHANGELOG.md

## 0.1.0 (2023-09-15)

Features:

  - Initial commit

## 0.2.0 (2023-09-15)

Features:

  - Docs/Readme updates

## 0.3.0 (2023-09-16)

Features:

  - Allow arbitrary terms as pattern IDs


test/test_helper.exs
ExUnit.start(exclude: [benchmark: true])


test/benchmark_test.exs
defmodule BenchmarkTest do
  use ExUnit.Case
  @patterns for _ <- 1..1000, do: :crypto.strong_rand_bytes(20) |> Base.encode16()
  @patterns_with_id Enum.with_index(@patterns)
  @patterns_charlist Enum.map(@patterns, &to_charlist/1)

  @text Enum.reduce(
          @patterns,
          "",
          &(&2 <> " #{&1} #{:crypto.strong_rand_bytes(300) |> Base.encode16()}")
        )

  @short_text Enum.reduce(
                @patterns |> Enum.take(5),
                "",
                &(&2 <> " #{&1} #{:crypto.strong_rand_bytes(30) |> Base.encode16()}")
              )

  @text_charlist to_charlist(@text)
  @short_text_charlist to_charlist(@short_text)

  @tag timeout: :infinity
  @tag :benchmark
  test "speed" do
    {:ok, left_long_tree} = AhoCorasearch.build_tree(@patterns_with_id)
    {:ok, left_long_uniq_tree} = AhoCorasearch.build_tree(@patterns_with_id, unique: true)
    {:ok, standard_tree} = AhoCorasearch.build_tree(@patterns_with_id, match_kind: :standard)

    {:ok, standard_uniq_tree} =
      AhoCorasearch.build_tree(@patterns_with_id, match_kind: :standard, unique: true)

    {:ok, left_first_tree} =
      AhoCorasearch.build_tree(@patterns_with_id, match_type: :leftmost_first)

    {:ok, left_first_uniq_tree} =
      AhoCorasearch.build_tree(@patterns_with_id, match_type: :leftmost_first, unique: true)

    sick = ExAhoCorasick.new(@patterns)
    erl = :aho_corasick.build_tree(@patterns_charlist)

    IO.puts(
      "\n\nTesting build_tree with #{@patterns_with_id |> length} patterns(this lib vs others):\n"
    )

    Benchee.run(
      %{
        aho_corasearch_build_tree: fn ->
          {:ok, _} = AhoCorasearch.build_tree(@patterns_with_id)
        end,
        aho_corasick_elixir_build_tree: fn ->
          %ExAhoCorasick{} = ExAhoCorasick.new(@patterns)
        end,
        aho_corasick_erlang_build_tree: fn ->
          {_, _, _} = :aho_corasick.build_tree(@patterns_charlist)
        end
      },
      warmup: 0,
      time: 5
    )

    IO.puts(
      "\n\nTesting search(this lib vs others), long text. byte_size: #{byte_size(@text)}:\n"
    )

    Benchee.run(
      %{
        aho_corasearch_search: fn ->
          [_ | _] = AhoCorasearch.search(standard_uniq_tree, @text)
        end,
        aho_corasick_elixir_search: fn ->
          %MapSet{} = ExAhoCorasick.search(sick, @text)
        end,
        aho_corasick_erlang_search: fn ->
          [_ | _] = :aho_corasick.match(@text_charlist, erl)
        end
      },
      warmup: 0,
      time: 5
    )

    IO.puts(
      "\n\nTesting search lib vs other erlang/elixir libs, short text. byte_size: #{byte_size(@short_text)}:\n"
    )

    Benchee.run(
      %{
        aho_corasearch_search: fn ->
          [_ | _] = AhoCorasearch.search(standard_uniq_tree, @short_text)
        end,
        aho_corasick_elixir_search: fn ->
          %MapSet{} = ExAhoCorasick.search(sick, @short_text)
        end,
        aho_corasick_erlang_search: fn ->
          [_ | _] = :aho_corasick.match(@short_text_charlist, erl)
        end
      },
      warmup: 0,
      time: 5
    )

    IO.puts("\n\nTesting search libs different match kinds + unique/non-unique:\n")

    Benchee.run(
      %{
        aho_corasearch_left_long: fn ->
          [_ | _] = AhoCorasearch.search(left_long_tree, @short_text)
        end,
        aho_corasearch_left_long_uniq: fn ->
          [_ | _] = AhoCorasearch.search(left_long_uniq_tree, @short_text)
        end,
        aho_corasearch_left_first: fn ->
          [_ | _] = AhoCorasearch.search(left_first_tree, @short_text)
        end,
        aho_corasearch_left_first_uniq: fn ->
          [_ | _] = AhoCorasearch.search(left_first_uniq_tree, @short_text)
        end,
        aho_corasearch_standard: fn ->
          [_ | _] = AhoCorasearch.search(standard_tree, @short_text)
        end,
        aho_corasearch_standard_uniq: fn ->
          [_ | _] = AhoCorasearch.search(standard_uniq_tree, @short_text)
        end,
        aho_corasearch_standard_overlap: fn ->
          [_ | _] = AhoCorasearch.search(standard_tree, @short_text, overlap: true)
        end,
        aho_corasearch_standard_uniq_overlap: fn ->
          [_ | _] = AhoCorasearch.search(standard_uniq_tree, @short_text, overlap: true)
        end
      },
      warmup: 0,
      time: 5
    )

    IO.puts("\n\nTesting AhoCorasearch.search/3 vs AhoCorasearch.word_search/3 short text:\n")

    Benchee.run(
      %{
        aho_corasearch_left_long: fn ->
          [_ | _] = AhoCorasearch.search(left_long_tree, @short_text)
        end,
        aho_corasearch_left_long_word_search: fn ->
          [_ | _] = AhoCorasearch.word_search(left_long_tree, @short_text)
        end,
        aho_corasearch_standard: fn ->
          [_ | _] = AhoCorasearch.search(standard_tree, @short_text)
        end,
        aho_corasearch_standard_word_search: fn ->
          [_ | _] = AhoCorasearch.word_search(standard_tree, @short_text)
        end
      },
      warmup: 0,
      time: 5
    )

    IO.puts("\n\nTesting AhoCorasearch.search/3 vs AhoCorasearch.word_search/3 long text:\n")

    Benchee.run(
      %{
        aho_corasearch_left_long_long_text: fn ->
          [_ | _] = AhoCorasearch.search(left_long_tree, @text)
        end,
        aho_corasearch_word_search_left_long_long_text: fn ->
          [_ | _] = AhoCorasearch.word_search(left_long_tree, @text)
        end,
        aho_corasearch_standard_long_text: fn ->
          [_ | _] = AhoCorasearch.search(standard_tree, @text)
        end,
        aho_corasearch_standard_word_search_long_text: fn ->
          [_ | _] = AhoCorasearch.word_search(standard_tree, @text)
        end
      },
      warmup: 0,
      time: 5
    )

    alphabet =
      ~w(a b c d e f g h i j k l m n o p q r s t u v w x y z)
      |> Enum.flat_map(fn i -> [i, i <> i] end)

    patterns = alphabet |> Enum.with_index()
    ratio = floor(byte_size(@text) / byte_size(Enum.join(alphabet)))
    text = String.duplicate(Enum.join(alphabet), ratio)

    {:ok, best_case_tree} =
      AhoCorasearch.build_tree(patterns, unique: true, insensitive: false, match_kind: :standard)

    {:ok, worst_case_tree} = AhoCorasearch.build_tree(patterns, match_kind: :standard)

    IO.puts(
      "\n\nTesting AhoCorasearch best case vs worst case: long text(byte_size: #{byte_size(text)}, num_matches: #{AhoCorasearch.search(best_case_tree, text, overlap: true) |> length}):\n"
    )

    Benchee.run(
      %{
        aho_corasearch_worst_case_long_text: fn ->
          [_ | _] = AhoCorasearch.search(worst_case_tree, text, overlap: true)
        end,
        aho_corasearch_best_case_long_text: fn ->
          [_ | _] = AhoCorasearch.search(best_case_tree, text, overlap: true)
        end
      },
      warmup: 0,
      time: 5
    )

    text = String.slice(text, 0, 100)

    IO.puts(
      "\n\nTesting AhoCorasearch best case vs worst case: short text(byte_size: #{byte_size(text)}, num_matches: #{AhoCorasearch.search(best_case_tree, text, overlap: true) |> length}):\n"
    )

    Benchee.run(
      %{
        aho_corasearch_worst_case_short_text: fn ->
          [_ | _] = AhoCorasearch.search(worst_case_tree, text, overlap: true)
        end,
        aho_corasearch_best_case_short_text: fn ->
          [_ | _] = AhoCorasearch.search(best_case_tree, text, overlap: true)
        end
      },
      warmup: 0,
      time: 5
    )
  end
end


test/aho_corasearch_test.exs
defmodule AhoCorasearchTest do
  use ExUnit.Case
  alias AhoCorasearch.Tree

  @uniq_pats [{"abc", 1}, {"def", 4}, {"ghi", 7}]
  @dup_pats [{"abc", 1}, {"abc", 2}, {"def", 4}, {"def", 5}, {"ghi", 7}]
  @non_int_pats [{"abc", :abc}, {"def", :def}, {"ghi", :ghi}]
  @non_int_dup_pats [{"abc", :abc}, {"abc", :abc2}, {"def", :def}, {"ghi", :ghi}]

  describe "build_tree" do
    test "it correctly builds a tree" do
      assert {:ok, %Tree{unique: true, integer_ids: true, insensitive: true, map: nil}} =
               AhoCorasearch.build_tree(@uniq_pats, unique: true)

      assert {:ok, %Tree{unique: true, integer_ids: true, insensitive: false, map: nil}} =
               AhoCorasearch.build_tree(@uniq_pats, unique: true, insensitive: false)

      assert {:ok, %Tree{unique: true, integer_ids: false, insensitive: true, map: %{}}} =
               AhoCorasearch.build_tree(@non_int_pats, unique: true)

      assert {:ok, %Tree{unique: false, integer_ids: false, insensitive: false, map: %{}}} =
               AhoCorasearch.build_tree(@non_int_dup_pats, insensitive: false)
    end

    test "it handles duplicate strings" do
      {:ok, %Tree{unique: false, insensitive: true, map: map}} =
        AhoCorasearch.build_tree(@dup_pats)

      assert %{0 => [1, 2], 1 => [4, 5], 2 => [7]} == map
    end

    test "it validates tree uniqueness" do
      {:error, _} = AhoCorasearch.build_tree(@dup_pats, unique: true)
      {:error, _} = AhoCorasearch.build_tree(@non_int_dup_pats, unique: true)
    end

    test "it works with non-int values" do
      assert {:ok, %{map: %{0 => [%{key: :a}, %{key: :a2}]}}} =
               AhoCorasearch.build_tree([{"a", %{key: :a}}, {"a", %{key: :a2}}])

      assert {:ok, %{map: %{0 => %{key: :a}, 1 => %{key: :b}}, integer_ids: false}} =
               AhoCorasearch.build_tree([{"a", %{key: :a}}, {"b", %{key: :b}}], unique: true)
    end
  end

  describe "search" do
    setup do
      {:ok, unique} = AhoCorasearch.build_tree(@uniq_pats, unique: true)
      {:ok, dup} = AhoCorasearch.build_tree(@dup_pats)
      {:ok, non_int} = AhoCorasearch.build_tree(@non_int_pats, unique: true)
      {:ok, non_int_dup} = AhoCorasearch.build_tree(@non_int_dup_pats)
      {:ok, unique: unique, dup: dup, non_int: non_int, non_int_dup: non_int_dup}
    end

    test "it returns the correct unique response", %{unique: tree, non_int: non_int} do
      resp = AhoCorasearch.search(tree, "abababcfedefedghi")
      assert [{4, 7, 1}, {9, 12, 4}, {14, 17, 7}] == resp
      resp = AhoCorasearch.search(non_int, "abababcfedefedghi")
      assert [{4, 7, :abc}, {9, 12, :def}, {14, 17, :ghi}] == resp
    end

    test "it returns the correct dup response", %{dup: tree, non_int_dup: non_int} do
      resp = AhoCorasearch.search(tree, "abababcfedefedghi")
      assert [{4, 7, [1, 2]}, {9, 12, [4, 5]}, {14, 17, [7]}] == resp
      resp = AhoCorasearch.search(non_int, "abababcfedefedghi")
      assert [{4, 7, [:abc, :abc2]}, {9, 12, [:def]}, {14, 17, [:ghi]}] == resp
    end

    test "it find the longest leftmost match" do
      patterns = [{"abcdefg", 1}, {"cdefghijkl", 2}, {"def", 4}, {"xyz", 5}]

      {:ok, tree} =
        AhoCorasearch.build_tree(patterns, unique: true, match_kind: :leftmost_longest)

      assert [{0, 7, 1}] ==
               AhoCorasearch.search(tree, "abcdefghijklmnop")
    end

    test "it works correctly for unicode" do
      patterns = [{"全世界", 3}, {"世界", 2}, {"に", 1}]
      {:ok, tree} = AhoCorasearch.build_tree(patterns, unique: true)
      assert [{0, 9, 3}, {12, 15, 1}] = AhoCorasearch.search(tree, "全世界中に")
    end
  end

  describe "word_search" do
    test "it only returns full word results" do
      patterns = ["red", "car", "red car"] |> Enum.with_index()
      {:ok, tree} = AhoCorasearch.build_tree(patterns, match_kind: :standard)
      results = AhoCorasearch.search(tree, "red car redcar", overlap: true, word_search: true)
      assert Enum.sort([{0, 3, [0]}, {0, 7, [2]}, {4, 7, [1]}]) == Enum.sort(results)
      results = AhoCorasearch.search(tree, "red car redcar", overlap: true, word_search: false)

      assert Enum.sort([{0, 3, [0]}, {0, 7, [2]}, {4, 7, [1]}, {8, 11, [0]}, {11, 14, [1]}]) ==
               Enum.sort(results)
    end
  end

  describe "insensitive" do
    test "setting to false" do
      {:ok, tree} = AhoCorasearch.build_tree([{"AAA", :aaa}, {"bbb", 1}], insensitive: false)
      assert [] == AhoCorasearch.search(tree, "aaaBBB")
      assert [{0, 3, [:aaa]}, {3, 6, [1]}] = AhoCorasearch.search(tree, "AAAbbb")
    end

    test "setting to true" do
      {:ok, tree} = AhoCorasearch.build_tree([{"AAA", :aaa}, {"bbb", 1}], insensitive: true)
      assert [{0, 3, [:aaa]}, {3, 6, [1]}] = AhoCorasearch.search(tree, "aaaBBB")
      assert [{0, 3, [:aaa]}, {3, 6, [1]}] = AhoCorasearch.search(tree, "AAAbbb")
    end
  end

  test "heap_bytes" do
    patterns =
      ~w(aaaaa bbbbbbb cccccccc dddddddd eeeeeee)
      |> List.duplicate(100)
      |> List.flatten()
      |> Enum.with_index()

    {:ok, tree} = AhoCorasearch.build_tree(patterns)
    heap_size = AhoCorasearch.heap_bytes(tree)
    assert heap_size > 0
  end
end


mix.exs
defmodule AhoCorasearch.MixProject do
  use Mix.Project

  def project do
    [
      app: :aho_corasearch,
      description:
        "Elixir lib for Aho-Corasick string searching. Uses a Rust-based NIF for greatly improved performance.",
      version: "0.3.0",
      elixir: "~> 1.15",
      name: "AhoCorasearch",
      source_url: "https://github.com/pejrich/AhoCorasearch",
      docs: [
        main: "AhoCorasearch",
        extras: ["Usage.md"]
      ],
      package: [
        name: "aho_corasearch",
        licenses: ["MIT"],
        links: %{"GitHub" => "https://github.com/pejrich/AhoCorasearch"},
        source_url: "https://github.com/pejrich/AhoCorasearch",
        files: ~w(lib priv native .formatter.exs mix.exs README* LICENSE*
                 CHANGELOG*)
      ],
      rustler_crates: rustler_crates(),
      deps: deps(),
      elixirc_paths: elixirc_paths(Mix.env())
    ]
  end

  defp elixirc_paths(:test), do: ["lib", "test/benchmark_libs"]
  defp elixirc_paths(_), do: ["lib"]

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger, :tools]
    ]
  end

  defp rustler_crates do
    [
      aho_corasearch_nif: [
        path: "native/aho_corasearch_nif",
        mode: :release
      ]
    ]
  end

  defp deps do
    [
      {:ex_doc, ">= 0.0.0", only: :dev, runtime: false},
      {:makeup_html, ">= 0.0.0", only: :dev, runtime: false},
      {:rustler, "~> 0.29"},
      {:benchee, "~> 1.0", only: [:test, :dev]},
      {:aho_corasick, git: "https://github.com/wudeng/aho-corasick", only: [:test]}
    ]
  end
end


lib/aho_corasearch/tree.ex
defmodule AhoCorasearch.Tree do
  @moduledoc false
  alias __MODULE__
  alias AhoCorasearch.{TreeBuilder, Native}

  defstruct [
    :ref,
    :tree_ref,
    :match_kind,
    integer_ids: false,
    insensitive: true,
    unique: false,
    map: %{}
  ]

  @type match_kind :: :standard | :leftmost_longest | :leftmost_first
  @type t :: %Tree{
          ref: binary(),
          tree_ref: reference(),
          match_kind: match_kind,
          insensitive: boolean(),
          integer_ids: boolean(),
          unique: boolean(),
          map: %{integer => term() | list(term())}
        }

  defdelegate new(patterns, opts), to: TreeBuilder

  def search(
        %Tree{
          map: map,
          tree_ref: tree,
          match_kind: mk,
          unique: unique,
          integer_ids: integer_ids,
          insensitive: insensitive
        },
        string,
        opts \\ [overlap: false, word_search: false]
      )
      when is_binary(string) do
    overlap = Keyword.get(opts, :overlap, false)
    word_search = Keyword.get(opts, :word_search, false)
    string = maybe_downcase(string, insensitive)

    do_search(tree, string, mk, overlap)
    |> fix_dups(map, unique: unique, int: integer_ids)
    |> maybe_filter(string, word_search)
  end

  def heap_bytes(%{tree_ref: tree}), do: Native.tree_heap_bytes(tree)

  defp maybe_filter(results, string, true), do: AhoCorasearch.WordSearch.filter(results, string)
  defp maybe_filter(results, _, _), do: results

  defp maybe_downcase(string, true), do: AhoCorasearch.Native.downcase(string)
  defp maybe_downcase(string, _), do: string

  defp fix_dups(res, _, unique: true, int: true), do: res

  defp fix_dups(res, map, unique: true, int: false),
    do: Enum.map(res, fn {s, e, id} -> {s, e, Map.get(map, id, id)} end)

  defp fix_dups(res, map, unique: false, int: false),
    do: Enum.map(res, fn {s, e, id} -> {s, e, Map.get(map, id, [id])} end)

  defp do_search(tree, string, :leftmost_longest, _), do: Native.leftmost_find_iter(tree, string)
  defp do_search(tree, string, :leftmost_first, _), do: Native.leftmost_find_iter(tree, string)
  defp do_search(tree, string, :standard, false), do: Native.find_iter(tree, string)
  defp do_search(tree, string, :standard, true), do: Native.find_overlapping_iter(tree, string)

  defimpl Inspect do
    import Inspect.Algebra

    def inspect(%{ref: ref, unique: uniq}, opts) do
      id = :erlang.ref_to_list(ref) |> Enum.drop(5) |> to_string()
      name = if uniq, do: "UniqueTree", else: "Tree"
      to_doc("#AhoCorasearch.#{name}<#{id}", opts)
    end
  end
end


lib/aho_corasearch/word_search.ex
defmodule AhoCorasearch.WordSearch do
  @moduledoc """
  WordSearch enables a more refined search mechanism that takes into account full words.

  For example "Earphones and pinecones" using the regular(faster) `AhoCorasearch.search/2` function will match(depending on settings and input): `["ear", "phone", "earphone", "and", "pine", "cone", "cones", "pinecone", "pinecones"]`. Which might be what you want. This module will help if what you'd expect to match is `["earphones", "and", "pinecones"]`(assuming all those words are in your tree).

  It's a slower search method than the regular search, by a reasonable factor.
  """
  def filter(results, string) do
    string
    |> word_boundary_indexes()
    # Remove any entry where either it's start / stop are not part of the boundary map
    |> then(
      &Enum.filter(results, fn {b, e, _} -> MapSet.member?(&1, b) && MapSet.member?(&1, e) end)
    )
  end

  @re :re.compile(~c"[[:punct:][:space:]]", [:unicode]) |> then(fn {:ok, re} -> re end)

  defp word_boundary_indexes(string) do
    # Find all indexes of non-word characters
    :re.run(string, @re, [:global, {:capture, :all, :index}])
    |> then(fn
      :nomatch -> []
      {:match, match} -> match
    end)
    # Regex.scan(~r/[[:punct:][:space:]]/u, string, return: :index)
    # Create mapset of start / end of each section of non-word characters
    |> Enum.reduce([0, byte_size(string)], fn [{start, length}], acc ->
      [start + length | [start | acc]]
    end)
    |> MapSet.new()
  end
end


lib/aho_corasearch.ex
defmodule AhoCorasearch do
  @moduledoc """
  Documentation for `AhoCorasearch`.
  """
  alias AhoCorasearch.Tree

  @type pattern :: binary
  @type start :: integer
  @type stop :: integer
  @type id :: term
  @type patterns :: [{pattern, id}]
  @type match :: {start, stop, id} | {start, stop, list(id)}
  @type matches :: list(match)
  @type tree :: Tree.t()

  @doc """
  Builds the tree needed for efficient string searching

  ## Examples
  ```elixir
  iex> AhoCorasearch.build_tree([{"abc", 1}, {"abc", 2}, {"def", 36}, {"xyz", 72}])
  {:ok, "#AhoCorasearch.Tree<0.1080113559.418906113.32320>"}

  iex> AhoCorasearch.build_tree([{"abc", 1}, {"def", 2}], unique: true, match_kind: :standard, insensitive: false)
  {:ok, "#AhoCorasearch.UniqueTree<0.1080113559.418906113.32348>"}

  iex> AhoCorasearch.build_tree([{"abc", 1}, {"abc", 2}, {"def", 36}, {"xyz", 72}], unique: true)
  {:error, :not_unique}
  ```

  The options it accepts are:
  * `unique` - Whether the keys in the pattern list are unique or not. This determines what the result tuple will look like. `{integer, integer, integer}` for unique trees, and `{integer, integer, list(integer)}` for non-unique trees. Defaults to `false`.
  * `insensitive` - If the matching should be case insensitive. Insensitivity is acheived by downcasing the patterns, and also downcasing the search text. This has a minor performance impact, but in most cases a negligible one. Defaults to `true`.
  * `match_kind` - There are 3 different matching options, but they must be chosen a tree building time, not at search time. The options are `standard`, `leftmost_longest`, or `leftmost_first`. Defaults to `standard`.

  Given the following patterns:
  ```elixir
  [
    {"headphone", 0},
    {"head", 1},
    {"phone", 2},
    {"pine", 3},
    {"cone", 4},
    {"pinecone", 5}
  ]
  ```
  Searching with `AhoCorasearch.search(tree, "headphones and pinecones")`

  * `standard`:
    * will return `[{0, 4, [1]}, {4, 9, [2]}, {15, 19, [3]}, {19, 23, [4]}]`
    * which translates to `["head", "phone", "pine", "cone"]`

  * `leftmost_longest`:
    * will return `[{0, 9, [0]}, {15, 23, [5]}]`
    * which translates to `["headphone", "pinecone"]`

  * `leftmost_first`:
    * will return `[{0, 9, [0]}, {15, 19, [3]}, {19, 23, [4]}]`
    * which translates to `["headphone", "pine", "cone"]`(notice that `first` here is first in the pattern list, not first in the search string).

  The examples above demonstrate the return value for a non-unique tree. The results is `{start_index, end_index, ids}`. Not that the indexes are in bytes, so you'll want to use `:binary.part/3` to extract substrings, not `String.slice/3`. The ID part must be an integer, but it does not need to be unique. Multiple keys can use the same ID. If multiple patterns are given with the same key and ID, the result will be repeated, for example: `[{34, 35, [1, 1]}]`. IDs are returned in the same order they are given, so for patterns: `[{"a", 3}, {"a", 1}, {"a", 2}, {"a", 1}]`, an "a" in the search text would give the result: `{X, Y, [3, 1, 2, 1]}`
  """
  @default_options [unique: false, insensitive: true, match_kind: :standard]
  @type match_kind :: :standard | :leftmost_longest | :leftmost_first
  @type options :: [unique: boolean, insensitive: boolean, match_kind: match_kind()]
  @spec build_tree(patterns, options) :: {:ok, tree} | {:error, term}
  def build_tree(patterns, opts \\ @default_options),
    do: Tree.new(patterns, opts)

  @doc """
  Searches the given string, against the given tree.

  This is one of two available ways to search. This is the faster of the two methods, but also might give more unexpected data. It will return matches anywhere they appear in the string(assuming the rules still abide by the tree's match configuration), but it has no notion of words. So partial and subwords will match. If you're looking for only full word matches, look at the `AhoCorasearch.word_search/3`, or use the option `word_search: true` which will filter out any matches that are not complete words(starting and ending at a word boundary).

  The inputs are:
  * `tree`: `AhoCorasearch.Tree.t()` - The tree that was build with `build_tree/2-3`
  * `string`: `String.t()` - The string that will be searched against the tree
  * `opts`:
    * `overlap: boolean`(default: false) - This argument is ONLY used if the tree type is `standard`. For trees with another `match_kind` it is ignored.
    * `word_search: boolean`(default: false) - Applies an additional filtering step to remove any matches that aren't complete words

  The result is a list of matches, either in the form `{integer, integer, integer}` for unique tree, or `{integer, integer, list(integer)}` for non-unique trees. The values are: `{start_index, end_index, ID}`. ID is whatever you passed in as the pattern(`[{"string", ID}, ...]`).

  The `start_index` and `end_index` are in BYTE values, not character or codepoints, so be sure you operate on the string with them by using `:binary.part/3` rather than `String.slice/3`.

  returns a list of all the matches in the input string.
  leftmost, longest matches
  While this handles unicode optimally, the return start/stop values are
  in bytes. i.e. use `:binary.part/3` rather than `String.slice/3`
  """
  @type search_opts :: [overlap: boolean, word_search: boolean]
  @spec search(Tree.t(), binary, search_opts) :: matches
  def search(%Tree{} = tree, string, opts \\ [overlap: false, word_search: false]),
    do: Tree.search(tree, string, opts)

  @doc """
  This is a convenience function for doing a word based search.

  Internally it just sets the `word_search` option to true for `search/3`, so either can be used as they're the same.
  """
  @spec search(Tree.t(), binary, overlap: boolean) :: matches
  def word_search(tree, string, opts \\ [overlap: false]),
    do: search(tree, string, Keyword.put(opts, :word_search, true))

  @doc """
  Returns the number of bytes that the tree is using on the native/Rust side.

  This does not account for any memory being stored on the Beam side, which hold the duplicate map
  Total size(beam + native) can be approximated with:
  `:erts_debug.flat_size(tree) + AhoCorasearch.heap_bytes(tree)`

  This function is only intended for use in development/debugging. I don't know the implications(performance or safety) of using it in production.
  """
  @spec heap_bytes(Tree.t()) :: pos_integer()
  def heap_bytes(%Tree{} = tree), do: Tree.heap_bytes(tree)
end
